include/psa/crypto_sizes.h

```diff
@@ -192,31 +192,31 @@
 
 /* The maximum size of an ECC key on this implementation, in bits.
  * This is a vendor-specific macro. */
-#if defined(MBEDTLS_ECP_DP_SECP521R1_ENABLED)
+#if defined(MBEDTLS_ECP_DP_SECP521R1_ENABLED) || defined(PSA_WANT_ECC_SECP_R1_521)
```

Comment:
```suggestion
#if defined(PSA_WANT_ECC_SECP_R1_521)
```
config_psa.h takes care of defining MBEDTLS_ECP_DP_SECP521R1_ENABLED if PSA_WANT_ECC_SECP_R1_521 is defined and vice versa.

Reply:
Isn't `MBEDTLS_ECP_DP_SECP521R1_ENABLED` re-defined if `MBEDTLS_PSA_ACCEL_ECC_SECP_R1_521` is not (i.e. if that curve is not accelerated)?
```
#if defined(PSA_WANT_ECC_SECP_R1_521)
#if !defined(MBEDTLS_PSA_ACCEL_ECC_SECP_R1_521)
```

The same hold for the others

Reply:
> Isn't `MBEDTLS_ECP_DP_SECP521R1_ENABLED` re-defined if `MBEDTLS_PSA_ACCEL_ECC_SECP_R1_521` is not (i.e. if that curve is not accelerated)?

If MBEDTLS_ECP_DP_SECP521R1_ENABLED is already defined in mbedtls_config.h, yes it is re-defined in config_psa.h if MBEDTLS_PSA_CRYPTO_CONFIG and PSA_WANT_ECC_SECP_R1_521 are enabled and MBEDTLS_PSA_ACCEL_ECC_SECP_R1_521 is not. As long as the definitions are the same the compilers do not complain.

Reply:
> But the curves are accelerated in this test... Am I missing something?

I am not sure what you mean here. To come back to my change suggestion, if the curves are accelerated it means that we are using the PSA configuration (MBEDTLS_PSA_CRYPTO_CONFIG is enabled) and thus the support for the curve is determined by the definition or not of PSA_WANT_ECC_SECP_R1_521.

Otherwise, if we are not using the PSA configuration (MBEDTLS_PSA_CRYPTO_CONFIG is disabled), the support for the curve is to start with determined by the definition or not of MBEDTLS_ECP_DP_SECP521R1_ENABLED in mbedtls_config.h and in config_psa.h we have:

#if defined(MBEDTLS_ECP_DP_SECP521R1_ENABLED)                                   
#define MBEDTLS_PSA_BUILTIN_ECC_SECP_R1_521 1                                   
#define PSA_WANT_ECC_SECP_R1_521                                                
#endif

Thus in all cases the curve ECC_SECP_R1_521 is supported if and only if PSA_WANT_ECC_SECP_R1_521 is defined and we do not need to care about MBEDTLS_ECP_DP_SECP521R1_ENABLED here.

Reply:
> We do not support an external PSA library

Ok, I think this is the bit of information I was missing. I thought that the purpose of `libtestdriver1.a` was to create a sort of generic PSA compliant driver library to interface to (and with "generic" I mean that I could be replaced with any other PSA compliant implementation).

> If MBEDTLS_ECP_DP_xxx is accessible through the PSA API, then crypto_config.h will take care of defining PSA_WANT_ECC_xxx

But having `MBEDTLS_ECP_DP_xxx` doesn't cause some extra code (for example from `ecp_curves.c`) to be taken into the build even though that curve is accelerated?

Reply:
The purpose of libtestdriver1 is to test the driver interface in the core (the part that isn't in libtestdriver1): we want to validate that when a driver for some mechanism is enabled, the core calls the driver (and not the built-in implementation).

If `MBEDTLS_ECP_DP_xxx` is enabled, that causes the relevant part of `ecp.c` and `ecp_curves.c` to be built (generic code + any curve-specific optimized code). If the curve is accelerated, and you don't need access to that curve from legacy functions that can't yet dispatch to PSA, then you don't need `MBEDTLS_ECP_DP_xxx` — assuming that Mbed TLS behaves correctly if a curve is only built as accelerated, which is what we're working on here.

Reply:
In addition to what Ronald said, I think it's important to note that here we're "in PSA context": this is a PSA header and we're defining a PSA macro. So as a general principle we should use only `PSA_WANT` macros. (This is a pre-existing issue, but since we're fixing this part of the code, let's do it the right way.)

Then, code that's doing ECC using the PSA APIs can just use this macro. Code that's doing ECC and might be using either PSA or legacy will need to check both `PSA_VENDOR_ECC_MAX_CURVE_BITS` and `MBEDTLS_ECP_MAX_BITS` and probably take the maxium or whatever, and that's OK.

But for pure-PSA code, things should remain clean.